controller/uninstall_controller.go

```diff
@@ -0,0 +1,329 @@
+	return c
+}
+
+func notifyNonBlocking(ch chan<- struct{}) {
+	select {
+	case ch <- struct{}{}:
+	default:
+	}
+}
+
+func notifyEventHandler(ch chan struct{}) cache.ResourceEventHandler {
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { notifyNonBlocking(ch) },
+		UpdateFunc: func(old, cur interface{}) { notifyNonBlocking(ch) },
+		DeleteFunc: func(obj interface{}) { notifyNonBlocking(ch) },
+	}
+}
+
+func (c *UninstallController) Uninstall(doneCh chan struct{}) error {
+	defer func() {
+		doneCh <- struct{}{}
+	}()
+	logrus.Infof("Uninstalling...")
+
+	if !controller.WaitForCacheSync("longhorn uninstall", doneCh,
+		c.volumeInformerSynced, c.engineImageInformerSynced,
+		c.nodeInformerSynced, c.daemonSetInformerSynced) {
+		return errors.New("Failed to sync informers")
+	}
+
+	if err := c.uninstall(); err != nil {
+		return err
+	}
+	logrus.Infof("Uninstallation complete")
+	return nil
+}
+
+func (c *UninstallController) uninstall() error {
+	if err := c.deleteCRDs(); err != nil {
+		return err
+	}
+	if running, err := c.isManagerRunning(); err != nil {
+		return err
+	} else if running {
+		if err := c.DeleteManager(); err != nil {
+			return err
+		}
+		// Re-run CRD deletion since manager may recreate things
+		// We might consider a Longhorn system "pause" to fix this race
+		if err := c.deleteCRDs(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *UninstallController) deleteCRDs() error {
+	if err := c.DeleteVolumes(); err != nil {
+		return err
+	}
+	if err := c.DeleteEngineImages(); err != nil {
+		return err
+	}
+	if err := c.DeleteNodes(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *UninstallController) DeleteVolumes() error {
+	vols, err := c.ds.ListVolumesRO()
+	if err != nil {
+		return err
+	}
+	if len(vols) == 0 {
+		return nil
+	}
+	volumesInUse := false
+	for _, vol := range vols {
+		logFields := logrus.Fields{
+			"type": "volume",
+		}
+		if vol.Status.State == types.VolumeStateAttaching ||
+			vol.Status.State == types.VolumeStateAttached {
+			logFields["name"] = vol.Name
+			logFields["state"] = vol.Status.State
+			logrus.WithFields(logFields).Warnf("In Use")
+			volumesInUse = true
+		}
+	}
+	if volumesInUse && !c.force {
+		logrus.Fatalf("Volume(s) in use. Detach all volumes or use --force to proceed anyway.")
+	}
+	for {
+		logFields := logrus.Fields{
+			"type": "volume",
+		}
+		vols, err = c.ds.ListVolumesRO()
+		if err != nil {
+			return err
+		}
+		if len(vols) == 0 {
+			logrus.WithFields(logFields).Infof("Removed all")
+			break
+		}
+		for _, vol := range vols {
+			logFields["name"] = vol.Name
+			timeout := metav1.NewTime(time.Now().Add(-GracePeriod))
+			if vol.DeletionTimestamp == nil {
+				if err := c.ds.DeleteVolume(vol.Name); err != nil {
+					logFields["error"] = err.Error()
+					logrus.WithFields(logFields).Warnf("Failed to mark for deletion")
+					notifyNonBlocking(c.volChange)
```

Comment:
Is it possible for us to lose events here when the channel buffer is full?

Reply:
Yes..refactor with queue imminent